### PR TITLE
Attempt to publish to at least mesh_n peers

### PR DESCRIPTION
**Pull request by AgeManning**
As we have now disabled flood-publish, we only publish to mesh peers. The mesh can be difficult to manage as  the management rules follow the gossipsub specification. 

Thanks to @pawanjay176 - We've noticed that it can be the case that we have connected peers on topics but these peers are not in our mesh (perhaps due to their own mesh requirements). Currently, we fail to publish the message if there are no peers in our mesh. 

This PR adjusts this logic to always attempt to publish to at least `mesh_n` peers. If we have peers that are subscribed to a topic, we will now attempt to publish messages to them (provided they have the required score). 

While I was here, I also separated the error types between insufficient peers and whether the queue is full, to help debugging and avoid any confusion in the future.

### CODE REVIEW

**Review by jimmygchen — COMMENTED** [review]
Nice, looks good to me 👍 
Looks like the `test_publish_without_flood_publishing` test needs a fix.

File: beacon_node/lighthouse_network/src/gossipsub/behaviour.rs [hunk 1]
```diff
@@ -635,9 +635,40 @@ where
...
+                        // Get a random set of peers that are appropriate to send messages too.
+                        let peer_list = get_random_peers(
+                            &self.connected_peers,
+                            &topic_hash,
+                            needed_extra_peers,
+                            |peer| {
+                                !mesh_peers.contains(peer)
+                                    && !explicit_peers.contains(peer)
+                                    && scores
+                                        .map(|(score, thresholds)| {
```
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/gossipsub/behaviour.rs]: maybe we can use the
```rust
!self
    .score_below_threshold(p, |pst| pst.publish_threshold)
    .0
```
pattern that's used everywhere else in the behaviour?

**Review by pawanjay176 — APPROVED** [review]
LGTM. Great fix 🎉

**Comment by AgeManning** [comment]
@mergifyio queue

**Pull request merged**